sunpy/coordinates/frames.py

```diff
@@ -345,8 +348,8 @@ def calculate_distance(self):
         rep = self.represent_as(UnitSphericalWrap180Representation)
         lat, lon = rep.lat, rep.lon
         alpha = np.arccos(np.cos(lat) * np.cos(lon)).to(lat.unit)
-        c = self.D0**2 - self.rsun**2
-        b = -2 * self.D0.to(u.m) * np.cos(alpha)
+        c = self.observer.radius**2 - self.rsun**2
```

Comment:
I don't think there is any reason anything needs casting to specific units, it's Quantity maths all the way through. (I have removed one of the explicit casts).